### ISSUE TITLE
Build Env: Docker vs DUE

### ISSUE THREAD

**Issue by tchandramohan** [CumulusNetworks/DUE#40]
Hi,
I have a basic insight of Docker and DUE as build environment for ONIE.
I want to know their pros and cons in order to make a decision.
-When and Why would you use DUE when you can achieve it on a simple docker?
-Does DUE require root access?
Thank you in advance.

**Comment by ehdoyle** [comment]
Hi. 
 The short answer is that using DUE is less work than setting up dockerfiles (the guy that wrote the template already figured out everything you need - why solve that problem again? ) and root access is not required.
 
Flippancy aside, the real answer is to use whatever set up works best for your needs. There are plenty of users that build ONIE without DUE, or I'm sure just Dockerfiles.  I'm partial to it because it has solved so many problems I kept running into setting up build environments on different systems, and it provides consistency  when users are spread out all over the globe.

For me, the real value add for DUE is that it handles so many limitations and corner cases that it just makes it easier for developers and end users to build software by allowing solved problems to stay solved.  That generalization is doing a lot of heavy lifting, but the end goal is to produce a specific build environment that looks a lot like what the user already has on the host system to let them get on with building code, and not be spending time tweaking their build environment.

To that end, the most noticeable thing with DUE is that it effectively transforms your system in to a known good build environment by adding the invoking user to the container for the software you're trying to build (ONIE, or Debian packages, or anything else a template has been written for.)  I know this seems like a very small thing but it has a HUGE impact on the end user experience.
Using ONIE as an example, you can write 4 docker files to build with Debian 8, Debian 9, Debian 10, and Debian 11, (depending on the age of the platform you're trying to build). Those will look mostly the same, but there will be some package renaming differences, and you'll need to make sure you add the code to pull in software that is not part of the regular build process.
Once you've got that container, you can just run a build command in it, but you'll have to  remember to mount the directory you want to build in when you start the container. And that's great if things build.
If you have to debug, then you'll want to log in to the container, where you've got whatever the default account was.

Or since a DUE  ONIE template exists, you can cut and paste an example from the due --create help, give it the version of Debian you want to build for and all the corner cases are handled because the problem was solved by the template writer.
With DUE, running the container can mount the build  directory, your home directory ( and any other directories, like the package cache under /var/cache/onie/downloads ) by default and it creates an account for you in the container so you still have access to your configuration files ( .gitconfig and .bashrc being a few nice ones ) and any files created in the container are saved with you as the owner.
Basically there are a  bunch of inconveniences that are always going to be  there if you're just using a dockerfile, but with DUE they're solved problems.
Also, by reducing the number of steps involved in setting up a build environment, it makes the software built from it easier to support as everybody is guaranteed to be using the same environment.
**Root access**
Due requires that the user be a member of the Docker group, but that's it for the host system. Another benefit of creating the user account in the container is that end users can do/install whatever they want without affecting the host system.
And if you're running Podman rather than Docker I don't think you even need that.

If you're running Debian an older version of DUE is in Debian 11 Buster, or you can install an updated .deb from GitHub.
The latest version has some updates, and that can be checked right out of github and run in place with makefiles, so if you are already using Docker/Podman, no install should be needed.
I think the documentation is pretty robust, but if you see things that aren't clear, do file a bug and we'll get that corrected quickly.
Thanks!

**Comment by tchandramohan** [comment]
Hi Doyle,

Thank you for your explanation. 

I Installed DUE within a container (Debian 10). 
I have two files in the directory -> due/due-build-merge/
1. onie-build-debian-10
2. onie-build-debian-10-template-merge

Now, from the due directory, I run the command
/home/due: ./due --create --from debian:10 --description "ONIE Build Debian 10" --name onie-build-debian-10 --prompt ONIE-10 --tag onie --use-template onie

Please refer to screenshot #1 
![image](https://user-images.githubusercontent.com/114584567/193789643-3ba6be75-4029-4de5-a780-414306899a22.png)

I am not sure how to proceed from here. 
Since the files already exist in due/due-build-merge, I went to the next step.

Next step was to launch a container from the due directory. I used the command,
/home/due: ./due --run --dockerarg --privileged

I get an error -> see screenshot #2 
![image](https://user-images.githubusercontent.com/114584567/193790506-ff8ded2f-be7a-4f54-baeb-74236b16ce5d.png)

I tried every fix on stackoverflow for this error, but could not understand to proceed. Please help. 
https://stackoverflow.com/questions/44678725/cannot-connect-to-the-docker-daemon-at-unix-var-run-docker-sock-is-the-docker

**Comment by ehdoyle** [comment]
Hi - thanks for the detailed description.
I think there may be two issues
1. DUE should be installed on the host system. If I'm understanding your setup correctly it looks like you're running it in a container, and that won't have Docker access unless the socket from the host system was also mounted in the container. 

2. Try a clean container build by running `due --create --clean` to clear out the merge directories. They're left around after a build for debug purposes, but you'll get the messages you're seeing when you try to rebuild the same container.

For the host system install, there are a couple of options.

1 - If it's Debian 11, there's an older upstreamed version of DUE that can be installed with `apt-get install due`
2 - If it's a Debian/Ubuntu host, the last release .deb can be used
[3.1.0](https://github.com/CumulusNetworks/DUE/releases/tag/v3.1.0) with
 `dpkg -i due_3.1.0-1.deb ; apt install --fix-broken`
 To install the package and use apt to resolve the dependencies a dpkg install won't look up.
 
3 -Or it can be installed from a git checkout ( to use the latest 3.1.0 release, check out  the 3.1.0 branch with, `git checkout v3.1.0` ) then  `make install`  will handle installing all the dependencies.

If your host system is running Red Hat, you'll have to go the `make install` route (RPM packaging has been on the to-do list for a while) the current master branch has better support for running DUE to create and run the Debian ONIE build environment, but it is at the start of the run up to release, so while everything seems to work it hasn't been as thoroughly tested as the last release. 

So to summarize, the steps would be:

1. Install DUE on the host system (so that Docker and other dependencies are there).
2. Run DUE to create the ONIE Debian 10 build container ( which should be the same `due --create` command you were running)
3. Use DUE to launch the container from the host system with `due --run`  and select the image from the menu (or if there is only one DUE created image, DUE will skip the menu and just run it)
 ...at which point you should see the Docker command that was used to run the container print out, and a message about container output that looks something like what is below.
 Let me know how it goes.
 
 ` ----------------------------------------------------------------------------
 ------ : Start and log in to [ 7bde4dfb8dde ] with this command:      -- 
   docker run \
     --platform amd64\
     --interactive\
     --tty\
     --rm\
     --volume /home/adoyle:/home/adoyle\
     --volume /var/cache/onie:/var/cache/onie\
     --volume /dev:/dev\
     --name adoyle_due-onie-build-debian-10_1664896575\
     --hostname due-onie-build-debian-10 7bde4dfb8dde /usr/local/bin/container-create-user.sh\
     --docker-group-id 134\
     --username adoyle\
     --userid 1000\
     --groupid 1000\
     --groupname adoyle

 ----------------------------------------------------------------------------
 ------ Host file systems are mounted as follows: ---------------------------
 -
 -   --volume  /home/adoyle:/home/adoyle 
 -   --volume /var/cache/onie:/var/cache/onie  
 -   --volume /dev:/dev 
 - 
 - You now have  [ 2 ] DUE containers running on beast.
 - User limit is [ 10 ]
 -
 ----------------------------------------------------------------------------
 -------- Type 'exit' to leave.  Docker Container Output Follows ------------

 ___________________________________________________________________________
|                                                                           |
| adoyle : Creating user UID [ 1000 ] GID [ 1000 ] Group [ adoyle ]
| adoyle : /home directory already exists. Not creating.
| Configuring  [ adoyle ] sudo,  --gid 1000 , shell [ /bin/bash ].
| Passwordless [ adoyle ].
|___________________________________________________________________________|
|                                                                           |
|               Appending container /etc/due-bashrc                         |
|___________________________________________________________________________|
|                                                                           |
| Container log in text follows:                                            |
|___________________________________________________________________________|

Linux due-onie-build-debian-10 5.15.0-46-generic #49~20.04.1-Ubuntu SMP Thu Aug 4 19:15:44 UTC 2022 x86_64

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.
`

**Comment by tchandramohan** [comment]
Hi Alex,

Thank you for your explanation. 
It is clear to me with DUE. 
It is not the right place but I am in the middle of a deadline. I would appreciate it if you replied here:

I have two questions for you:
[Q1] What is not clear to me is generating signing keys after DUE as per this tutorial (https://www.youtube.com/watch?v=evzkiiRRIvw)

I am trying to run the command:
build@09964100ebc3:~/sec/onie/build-config$: make MACHINE=kvm_x86_64 signing-keys-install
The command runs, generates some keys in the directory: onie/encryption/machines/kvm_x86_64/keys but terminates with an error. I am attaching a screenshot. I do not know how to proceed. 
![image](https://user-images.githubusercontent.com/114584567/194057376-9e135b5f-4d03-467c-a6eb-9244d31cfe0d.png)

[Q2]. I disabled secure boot feature by following the steps listed here (https://github.com/opencomputeproject/onie/issues/992)
I ran the following commands
$: cd onie/build-config
$: make -j4 MACHINE=kvm_x86_64 all
The screenshot is shown below. I was surprised that .iso files were not generated in the onie/build/images directory
![image](https://user-images.githubusercontent.com/114584567/194057836-4db2f5e7-3f0b-4961-b868-73ca3396c07a.png)

Later, I ran the command -> $: make -j4 MACHINE=kvm_x86_64 all demo
The same result. The .iso files were not generated
![image](https://user-images.githubusercontent.com/114584567/194058106-06e72bf9-7393-48b6-84a9-9db8f0fb5333.png)

**Comment by ehdoyle** [comment]
Hi 
This is moving more towards ONIE build issues, but given there is some contextual overlap, let's keep going :)
[Q1] Rootless dockers  - it will work without using  --privileged when running Docker, but things like creating and populating the virtual USB drive used to hold the signing keys will have to be done outside of the container where the user can loopback mount the filesystem.  It seems to me that should be possible within the container, but it's one of those things that's not been tracked down.

So for key creation,
make MACHINE=kvm_x86_64 signing-keys-install  does invoke the 'signing-keys-generate' target and should put keys under
under encryption/machines/kvm_x86_64/keys, even if it doesn't have permissions to create the virtual USB drive.

make MACHINE=kvm_x86_64 shim-self-sign

[Q2]  cert-to-efi-sig-list comes out of the efitools package. The Debian 10 environment builds that from source to get the 1.9.2 version rather than the default 1.8.1-1 version that comes with Debian 10.  I forget exactly why this was needed, but it's one of those things that was solved in DUE two years ago and I haven't had to worry about since.
As an aside, there is also a problem with one of the encryption tools where, if your absolute directory path is too long it will fail, but I think the script using it has an error message to suggest this unexpected cause of failure.

[Q3] I suspect the parallel build died somewhere above. Sometimes the build does not correctly resolve dependencies from a failed build and a `make distclean` and rebuilding is the best option. Especially if that hasn't been done since changing the makefile. I'd suggest running the make with a -j1 so that it dies on the exact failure, or repeatedly build in parallel to exhaust the builds that do work until  it returns the actual error.  

I did just double check the secure boot build and the build without secure boot enabled, and the steps are correct, so I'd guess the two issues were not having the right efitools for the secure build, and needing to run a `make distclean` after changing to a non-secure build?

**Comment by tchandramohan** [comment]
Hi Alex,

An update on [Q3]. I did run make distclean and make MACHINE=kvm_x86_64 all
I was working towards launching ONIE on my host system for a KVM_x86_64 target using Debian-10 docker as build environment (not DUE).
The project fails at the step when I am building ONIE for a particular target (kvm_x86_64)
command used is shown below:
build@4a690b14c02:~/src/onie/build-config$: make MACHINE=kvm_x86_64 all

The project tries to generate a table for multilib using the locale variables, which are set to POSIX inside the container for building ONIE
I modified the original dockerfile (given by ONIE github) that was used to build the container changing the variables to en_us as shown below:

ARG DEBIAN_FRONTEND=noninteractive
ARG lc_all=en_US.UTF-8
ARG lang=en_US.UTF-8
ARG language=en_US.UTF-8

Then I recreated the build-env and launched a new container.
And I have the same error (screenshots attached)

![image](https://user-images.githubusercontent.com/114584567/194324431-7bb4bd72-181f-471e-9f3b-ef7065968b73.png)
![image](https://user-images.githubusercontent.com/114584567/194324532-638c265a-76cd-46f6-8471-77e4efec0bbe.png)

**Comment by ehdoyle** [comment]
Looks like there was a resolution in ONIE. Closing out.
